### PR TITLE
improve rendering of waveform marks

### PR DESCRIPTION
**Pull request by m0dB**
This PR improves the rendering of the waveform marks, by auto-scaling and better aligning the symbolic marks (such as the intro/outro triangles and the loop open circle arrow), as discussed here: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/look.20of.20loop.20out.20marker

Additionally it:
- removes code duplication between `waveformrendermark.cpp` and `allshaders/waveformrendermark.cpp` by moving the mark image rendering to `waveformmark.cpp` and doing a general cleanup of that code
- doesn't draw a small empty rectangle for marks without a label
- fixes drawing of marks in vertical orientation in legacy waveforms
- fixes a small bug with hovering over marks (propagate leave event)
- uses antialiasing to render the marks and the glsl play position mark image

Screenshot 2023-10-21 at 03 34 09

### CODE REVIEW

**Comment by acolombier** [comment]
What happens when labels overlap? (e.g in the screenshot, label of hotcue 4 would be long enough to overlap label of 3)
Do you think it would be possible to add some logic to bring labels on the foreground? For example, the last activated on is always on foreground or if the user hover them on the waveform, or in the deck

**Comment by ronso0** [comment]
@acolombier I think you're referring to #11745 

I suggest to keep this PR compact, then take care of the other issues (#11745 and dark/bright detection / hover effect).

I'll test this asap.

**Comment by m0dB** [comment]
There is currently no such logic. But the case you describe (the previous marker being so long that it overlaps the next marker) could be very easily solved with very simple logic: draw the markers from right to left. 

But what annoys me most is when you have two markers at the same position (I actually use that quite a lot, e.g. a normal and a looping hotcue). I want to try placing the second one higher than the first in that case.

In any case, all that should be in a separate PR.

**Comment by acolombier** [comment]
@ronso0 this isn't quite the same now - my question was if instead of `OK!` as a label, you have `Some long text`(up to 23 chars + the `...` suffix), how will that be interfering with the 3rd hotcue? I assume one will overlap the other, and I'm wondering if we should account for further work (no particularly in that PR) to improve the UX

@m0dB I also use that behavior and suffering from the same limitation and the change you are describing would be very appreciated :)

**Comment by m0dB** [comment]
I think https://github.com/mixxxdj/mixxx/issues/11745 is more about the inconsistency between the scrolling waveform and the waveform overview.

File: src/waveform/renderers/waveformmark.h
Comment (Swiftb0y) [inline src/waveform/renderers/waveformmark.h:119]: so IIUC, `m_pTexture` is always `nullptr` when running with a legacy shader and QImage `isNull` the other way around, right? Wouldn't it be better instead then if these concepts lived in the shaders then instead? At least that sounds like thats a better separation of concern.

File: src/waveform/renderers/waveformmark.h
Comment (Swiftb0y) [inline src/waveform/renderers/waveformmark.h:120]: while I don't have any objection to this personally, the pre-established style was to always initialize all members only in the member initializer list of the constructor(s).

File: src/waveform/renderers/waveformmark.h
Comment (m0dB) [reply src/waveform/renderers/waveformmark.h:119]: In the renderer you mean? (WaveformMarkRenderer and allshaders::WaveformMarkRenderer). But that would require keeping track of which image or texture goes with which WaveformMark. No, this really needs to be part of WaveformMark. But maybe it can be done by subclassing WaveformMark... I will have a look, I agree this is not ideal.

File: src/waveform/renderers/waveformmark.h
Comment (m0dB) [reply src/waveform/renderers/waveformmark.h:120]: Ok, I wasn't aware of that. I was under the impression that the general push was to go for modern c++. Do you think this is open for discussion? Otherwise I will change it.

File: src/waveform/renderers/waveformmark.h
Comment (Swiftb0y) [reply src/waveform/renderers/waveformmark.h:119]: > But that would require keeping track of which image or texture goes with which WaveformMark.

The WaveformMarks in the renderer are just stored in a container, should just be sufficient to store a `std::pair<std::unique_ptr<QOpenGLTexture>, WaveformMark>` and then implement `generateMarkImage` in terms of that, right?

File: src/waveform/renderers/waveformmark.h
Comment (Swiftb0y) [reply src/waveform/renderers/waveformmark.h:120]: I'm in favor of it, but you'll have to discuss with "the keeper of the styleguide" @daschuer. Do you have any references that this style is preferred over ours in modern c++?

**Comment by JoergAtGithub** [comment]
On my Win11 system the green loop marker looks different than in your screenshot:
grafik

**Comment by m0dB** [comment]
> On my Win11 system the green loop marker looks different than in your screenshot:

My screenshot is HDPI so that will make a difference for sure, but indeed this looks worse than it should... Can you try changing, in waveformmark.cpp:

```
// And we scale the font size accordingly.
m_font.setPointSizeF(50.0 * ratio);
```

to

```
m_font.setPixelSize(19);
```

and see if that improves things? (This will make the intro/outro markers look weird, but that's expected)

**Comment by m0dB** [comment]
and without this PR ?

**Comment by JoergAtGithub** [comment]
Plain 2.4:
grafik

**Comment by m0dB** [comment]
Thanks for testing and following up. Well, that's a relieve! That also looks bad!

What if you paste this character ↻ in a text editor with font Open Sans (bold) at different sizes?

And does it improve things if in `waveformmark.cpp` instead of `painter.drawText(pos, label);` you use `painter.drawText(static_cast<int>(pos.x()), static_cast<int>(pos.y()), label);` ?

**Comment by JoergAtGithub** [comment]
This PR with `painter.drawText(static_cast<int>(pos.x()), static_cast<int>(pos.y()), label);`:
grafik

**Comment by m0dB** [comment]
My conclusion is that Windows does a pretty bad job at rendering this symbol. Wondering what it looks like on Linux.

Maybe we should use a png instead?

**Comment by ronso0** [comment]
On Ubuntu Studio 20.04 it looks okay @150%
image

@250%
image

**Comment by JoergAtGithub** [comment]
> Maybe we should use a png instead?

If possible, we should use svg vector graphics, as in the skins.

File: src/waveform/renderers/waveformmark.h
Comment (m0dB) [reply src/waveform/renderers/waveformmark.h:119]: No, they are stored in a WaveformMarkSet. But I can derive a WaveformMarkImage and a WaveformMarkTexture from WaveformMark and templatize WaveformMarkSet.

Comment (m0dB) [reply src/waveform/renderers/waveformmark.h:119]: I started doing this, but it adds way too much complexity. I prefer having these two data members, of which either one is used depending on the context, instead of adding all this complexity.

**Comment by m0dB** [comment]
I have added the possibility to specify an svg.

This is with an svg for the loop mark.

Screenshot 2023-10-28 at 12 24 03

Does this look okay on non-HDPI, @JoergAtGithub ? Maybe the svg needs a bit of tweaking.

File: src/waveform/renderers/waveformmark.h
Comment (Swiftb0y) [reply src/waveform/renderers/waveformmark.h:120]: okay. Have you considered putting the members in a variant to highlight the fact that they are mutual exclusive? If that introduces too much complexity again, then ignore it. I'm just trying to come up with solutions to make the code more self-documenting.

Comment (m0dB) [reply src/waveform/renderers/waveformmark.h:120]: yes, that makes sense! will do!

Comment (m0dB) [reply src/waveform/renderers/waveformmark.h]: done. hope this is in line with what you had in mind.

File: src/waveform/renderers/waveformmark.h
Comment (m0dB) [reply src/waveform/renderers/waveformmark.h]: not relevant anymore

**Comment by JoergAtGithub** [comment]
Yes, the loop marker looks now fine on Windows too!
grafik
The Intro/Outro markers look not that bad as the loop markers looked before, but also there SVG should be used.

**Comment by m0dB** [comment]
> Yes, the loop marker looks now fine on Windows too!

Great!

> The Intro/Outro markers look not that bad as the loop markers looked before, but also there SVG should be used. Which can be done in another PR.

No, I will add it to this PR, it's very little effort.

**Comment by ronso0** [comment]
> No, I will add it to this PR, it's very little effort.

All skins? :grimacing: 
Reason for asking: once upon a time we decided to add the Sync Leader control to all skins (even Shade?), and just today I noticed I only managed to add it to LateNight.
It's an entirely different story (in terms of xml integration and design effort), though I wanted to add that note.

**Comment by m0dB** [comment]
That's a good point, @ronso0 . I only added this to Latenight, but the text rendering is still there, so of course other skins can continue using that.

**Comment by JoergAtGithub** [comment]
Now all the markers look fine on Windows11! Thank you!

File: src/waveform/renderers/waveformmark.h
Comment (Swiftb0y) [reply src/waveform/renderers/waveformmark.h]: Yes, almost. I was rather thinking of `std::variant` but this also works. The only thing I'm slighly concerned about is the blind static_cast from the parent to the derived type but I don't know any better solution either right now without introducing lots of boilerplate.

Comment (Swiftb0y) [reply src/waveform/renderers/waveformmark.h]: Thank you for going through the extra effort.

**Review by Swiftb0y — APPROVED** [review]
LGTM. thanks a lot. before merging, did someone test this on linux as well?

**Comment by ywwg** [comment]
looks good on linux

File: src/waveform/renderers/waveformmark.cpp
Comment (ywwg) [inline src/waveform/renderers/waveformmark.cpp:148]: technically this is a struct not a class, so let's call it that

File: src/waveform/renderers/waveformmark.cpp
Comment (ywwg) [inline src/waveform/renderers/waveformmark.cpp:182]: why is this scoping needed?

File: src/waveform/renderers/waveformmark.cpp
Comment (ywwg) [inline src/waveform/renderers/waveformmark.cpp:267]: please remove commented-out lines like this throughout

File: src/waveform/renderers/waveformmark.cpp
Comment (ywwg) [inline src/waveform/renderers/waveformmark.cpp:271]: unnecessary newline.  if possible, please run the code formatter on changed lines

File: src/waveform/renderers/waveformmark.cpp
Comment (ywwg) [reply src/waveform/renderers/waveformmark.cpp:182]: oh is this copied from the original code? then maybe that's fine 👍🏻

File: src/waveform/renderers/waveformmark.h
Comment (ywwg) [inline src/waveform/renderers/waveformmark.h:29]: I think this is equivalent and simpler, though we might have a style reason not to use this format? 
```suggestion
        bool m_obsolete = false;
```

File: src/waveform/renderers/waveformmark.h
Comment (ywwg) [inline src/waveform/renderers/waveformmark.h:115]: I'm not sure we use this style of initialization elsewhere, probably clearer to do = 0; instead

**Review by ywwg — COMMENTED** [review]
looks good! Please clean up stray commented-out lines before merging, and I've made some comments.  I am not blocking this PR on them though since I am famously bad at getting back to my reviews :)

File: res/skins/LateNight/waveform.xml
Comment (ronso0) [reply res/skins/LateNight/waveform.xml:65]: not sure about this (here). sometimes it's helpful to have the matching unicode char at hand.

Comment (m0dB) [reply res/skins/LateNight/waveform.xml:65]: Yes, I agree, the Text should be fallback. I will revert the commenting.

File: src/waveform/renderers/waveformmark.h
Comment (m0dB) [reply src/waveform/renderers/waveformmark.h:115]: I would like to get developer concensus on this. @Swiftb0y (IIRC) recently told me to initialize in the initializer list...

But there are two issues at hand here: 1) using curly braces aka uniform initialization and 2) in-class initialization vs initializer lists.

As for 1, from what I understand, uniform initialization is the prefered way since c++11. The standard https://isocpp.org/wiki/faq/cpp11-language#uniform-init clearly says: "Prefer initializing using {}."

As for 2, I don't think there is such a clear preference. When you have multiple constructors, in-class initialization certainly makes things easier, otherwise I don't think it matters much.

File: src/waveform/renderers/waveformmark.h
Comment (m0dB) [reply src/waveform/renderers/waveformmark.h:29]: See my comment below on uniform initialization.

File: src/waveform/renderers/waveformmark.h
Comment (Swiftb0y) [reply src/waveform/renderers/waveformmark.h:115]: > I would like to get developer concensus on this. @Swiftb0y (IIRC) recently told me to initialize in the initializer list...

Let me be clear, I didn't mean to tell you to use initializer lists, I just said other people will complain and I thought I'd mention it so it doesn't seem like I didn't review your PR thoroughly. I also prefer this style and if other devs do so too, then lets use it. If others complain we can settle this in a simple vote. 

> I'm not sure we use this style of initialization elsewhere, probably clearer to do = 0; instead


I also prefer `= 0` but only when the type is clearly numerical. On the other hand the meaning of `= 0` is already overloaded in the context of pure virtual member functions. So my preference is probably wrong and we should prefer using `{}` as per the cpp core guidelines.

File: res/skins/LateNight/waveform.xml
Comment (m0dB) [reply res/skins/LateNight/waveform.xml:65]: Done

File: src/waveform/renderers/waveformmark.cpp
Comment (m0dB) [reply src/waveform/renderers/waveformmark.cpp:148]: Done

File: src/waveform/renderers/waveformmark.cpp
Comment (m0dB) [reply src/waveform/renderers/waveformmark.cpp:182]: i removed it, as it allows capHeight to be const.

File: src/waveform/renderers/waveformmark.cpp
Comment (m0dB) [reply src/waveform/renderers/waveformmark.cpp:267]: Done

File: src/waveform/renderers/waveformmark.cpp
Comment (m0dB) [reply src/waveform/renderers/waveformmark.cpp:271]: I am using pre-commit , not sure why this wasn't formatted. Anyway, fixed.

File: src/waveform/renderers/waveformmark.h
Comment (m0dB) [reply src/waveform/renderers/waveformmark.h:115]: Ok, let's take this discussion to zulip

Comment (m0dB) [reply src/waveform/renderers/waveformmark.h]: removing in-class initialization

**Comment by m0dB** [comment]
All comments have been addressed, so I think this can be merged.

**Comment by JoergAtGithub** [comment]
Just tested this again, and it's still working. LGTM! Thank you!

**Pull request merged**
